File: paimon-format/src/main/java/org/apache/paimon/format/blob/BlobFormatReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.format.blob;

import org.apache.paimon.data.Blob;
import org.apache.paimon.data.BlobArrayPlaceholder;
import org.apache.paimon.data.BlobDescriptor;
import org.apache.paimon.data.BlobPlaceholder;
import org.apache.paimon.data.BlobRef;
import org.apache.paimon.data.GenericArray;
import org.apache.paimon.data.GenericRow;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.fs.FileIO;
import org.apache.paimon.fs.Path;
import org.apache.paimon.fs.SeekableInputStream;
import org.apache.paimon.reader.FileRecordIterator;
import org.apache.paimon.reader.FileRecordReader;
import org.apache.paimon.types.DataType;
import org.apache.paimon.types.DataTypeRoot;
import org.apache.paimon.utils.DeltaVarintCompressor;
import org.apache.paimon.utils.IOUtils;
import org.apache.paimon.utils.UriReader;

import javax.annotation.Nullable;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/** {@link FileRecordReader} for blob file. */
public class BlobFormatReader implements FileRecordReader<InternalRow> {

    private static final int ARRAY_HEADER_LENGTH = 9;
    private static final int ARRAY_INDEX_LENGTH_SIZE = Integer.BYTES;
    private static final int MIN_ARRAY_PAYLOAD_LENGTH =
            ARRAY_HEADER_LENGTH + ARRAY_INDEX_LENGTH_SIZE;

    // Shared by all BlobRefs this reader produces, so a BlobFormatWriter can reuse one source
    // stream.
    private final UriReader uriReader;
    private final Path filePath;
    private final String filePathString;
    private final BlobFileMeta fileMeta;
    private final @Nullable SeekableInputStream in;
    private final int fieldCount;
    private final int blobIndex;
    private final DataType blobFieldType;
    private final Object blobPlaceholder;
    private final boolean blobAsDescriptor;

    private boolean returned;

    public BlobFormatReader(
            FileIO fileIO,
            Path filePath,
            BlobFileMeta fileMeta,
            @Nullable SeekableInputStream in,
            int fieldCount,
            int blobIndex,
            DataType blobFieldType,
            boolean blobAsDescriptor) {
        this.uriReader = UriReader.fromFile(fileIO);
        this.filePath = filePath;
        this.filePathString = filePath.toString();
        this.fileMeta = fileMeta;
        this.in = in;
        this.fieldCount = fieldCount;
        this.blobIndex = blobIndex;
        this.blobFieldType = blobFieldType;
        this.blobPlaceholder =
                blobFieldType.getTypeRoot() == DataTypeRoot.ARRAY
                        ? BlobArrayPlaceholder.INSTANCE
                        : BlobPlaceholder.INSTANCE;
        this.blobAsDescriptor = blobAsDescriptor;
        this.returned = false;
    }

    @Nullable
    @Override
    public FileRecordIterator<InternalRow> readBatch() throws IOException {
        if (returned) {
            return null;
        }

        returned = true;
        return new FileRecordIterator<InternalRow>() {

            int currentPosition = 0;

            @Override
            public long returnedPosition() {
                return fileMeta.returnedPosition(currentPosition);
            }

            @Override
            public Path filePath() {
                return filePath;
            }

            @Nullable
            @Override
            public InternalRow next() {
                if (currentPosition >= fileMeta.recordNumber()) {
                    return null;
                }

                Object field;
                if (fileMeta.isNull(currentPosition)) {
                    field = null;
                } else if (fileMeta.isPlaceHolder(currentPosition)) {
                    field = blobPlaceholder;
                } else {
                    long offset = fileMeta.blobOffset(currentPosition) + 4;
                    long length = fileMeta.blobLength(currentPosition) - 16;
                    if (blobFieldType.getTypeRoot() == DataTypeRoot.ARRAY) {
                        field = readBlobArray(in, offset, length);
                    } else {
                        field = readBlob(in, offset, length);
                    }
                }
                currentPosition++;
                GenericRow row = new GenericRow(fieldCount);
                row.setField(blobIndex, field);
                return row;
            }

            @Override
            public void releaseBatch() {}
        };
    }

    private Blob readBlob(@Nullable SeekableInputStream in, long position, long length) {
        if (in != null && !blobAsDescriptor) {
            return Blob.fromData(readInlineBlob(in, position, length));
        }
        return new BlobRef(uriReader, new BlobDescriptor(filePathString, position, length));
    }

    @Override
    public void close() throws IOException {
        IOUtils.closeQuietly(in);
    }

    private static byte[] readInlineBlob(SeekableInputStream in, long position, long length) {
        byte[] blobData = new byte[(int) length];
        try {
            in.seek(position);
            IOUtils.readFully(in, blobData);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return blobData;
    }

    private GenericArray readBlobArray(SeekableInputStream in, long position, long length) {
        if (in == null) {
            throw new IllegalStateException("Input stream must be available for ARRAY<BLOB>.");
        }
        if (position < 0
                || length < MIN_ARRAY_PAYLOAD_LENGTH
                || position > Long.MAX_VALUE - length) {
            throw new IllegalArgumentException(
                    "Invalid ARRAY<BLOB> payload position or length: " + position + ", " + length);
        }

        try {
            byte[] header = new byte[ARRAY_HEADER_LENGTH];
            in.seek(position);
            IOUtils.readFully(in, header);
            ByteBuffer headerBuffer = littleEndianBuffer(header);
            int magic = headerBuffer.getInt();
            if (magic != BlobFormatWriter.ARRAY_MAGIC_NUMBER) {
                throw new IllegalArgumentException(
                        "Invalid ARRAY<BLOB> payload magic number: " + magic);
            }
            byte version = headerBuffer.get();
            if (version != BlobFormatWriter.ARRAY_VERSION) {
                throw new UnsupportedOperationException(
                        "Unsupported ARRAY<BLOB> payload version: " + version);
            }
            int elementCount = headerBuffer.getInt();
            if (elementCount < 0) {
                throw new IllegalArgumentException(
                        "Invalid ARRAY<BLOB> element count: " + elementCount);
            }

            long payloadEnd = position + length;
            long elementDataStart = position + ARRAY_HEADER_LENGTH;
            long elementIndexLengthPosition = payloadEnd - ARRAY_INDEX_LENGTH_SIZE;
            byte[] indexLengthBytes = new byte[ARRAY_INDEX_LENGTH_SIZE];
            in.seek(elementIndexLengthPosition);
            IOUtils.readFully(in, indexLengthBytes);
            int elementIndexLength = littleEndianBuffer(indexLengthBytes).getInt();
            long maximumIndexLength = length - MIN_ARRAY_PAYLOAD_LENGTH;
            if (elementIndexLength < 0 || elementIndexLength > maximumIndexLength) {
                throw new IllegalArgumentException(
                        "Invalid ARRAY<BLOB> element index length: " + elementIndexLength);
            }
            if (elementCount > elementIndexLength) {
                throw new IllegalArgumentException(
                        "ARRAY<BLOB> element count exceeds element index length.");
            }

            byte[] elementIndexBytes = new byte[elementIndexLength];
            long elementIndexStart = elementIndexLengthPosition - elementIndexLength;
            in.seek(elementIndexStart);
            IOUtils.readFully(in, elementIndexBytes);
            long[] elementLengths;
            try {
                elementLengths = DeltaVarintCompressor.decompress(elementIndexBytes);
            } catch (RuntimeException e) {
                throw new IllegalArgumentException("Invalid ARRAY<BLOB> element index.", e);
            }
            if (elementLengths.length != elementCount) {
                throw new IllegalArgumentException(
                        "ARRAY<BLOB> element count does not match element index length.");
            }

            long elementDataLength = elementIndexStart - elementDataStart;
            long totalElementLength = 0;
            for (long elementLength : elementLengths) {
                if (elementLength == BlobFormatWriter.ARRAY_NULL_ELEMENT_LENGTH) {
                    continue;
                }
                if (elementLength < 0) {
                    throw new IllegalArgumentException(
                            "Invalid ARRAY<BLOB> element length: " + elementLength);
                }
                if (!blobAsDescriptor && elementLength > Integer.MAX_VALUE) {
                    throw new IllegalArgumentException(
                            "ARRAY<BLOB> inline element is too large: " + elementLength);
                }
                if (elementLength > elementDataLength - totalElementLength) {
                    throw new IllegalArgumentException(
                            "ARRAY<BLOB> element lengths exceed the payload data length.");
                }
                totalElementLength += elementLength;
            }
            if (totalElementLength != elementDataLength) {
                throw new IllegalArgumentException(
                        "ARRAY<BLOB> element lengths do not match the payload data length.");
            }

            Object[] blobs = new Object[elementCount];
            long elementOffset = elementDataStart;
            for (int i = 0; i < elementCount; i++) {
                long elementLength = elementLengths[i];
                if (elementLength == BlobFormatWriter.ARRAY_NULL_ELEMENT_LENGTH) {
                    blobs[i] = null;
                } else if (blobAsDescriptor) {
                    blobs[i] =
                            new BlobRef(
                                    uriReader,
                                    new BlobDescriptor(
                                            filePathString, elementOffset, elementLength));
                    elementOffset += elementLength;
                } else {
                    blobs[i] = Blob.fromData(readInlineBlob(in, elementOffset, elementLength));
                    elementOffset += elementLength;
                }
            }

            return new GenericArray(blobs);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private static ByteBuffer littleEndianBuffer(byte[] bytes) {
        return ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
    }
}

File: paimon-python/pypaimon/format/blob/blob_format_reader.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

import struct
from typing import List, Optional, Iterator

from pypaimon.table.row.blob import BlobRef, BlobDescriptor
from pypaimon.table.row.generic_row import GenericRow
from pypaimon.format.blob.delta_varint_compressor import DeltaVarintCompressor


class BlobFormatReader:
    """
    Python implementation of Paimon's BlobFormatReader.

    Reads blob data from files in the Paimon blob format, which includes:
    - Reading the footer to get blob index information
    - Creating blob references for each blob entry
    - Supporting selection of specific blob entries
    """

    def __init__(self, file_path: str, file_size: int, selection: Optional[List[int]] = None):
        """
        Initialize BlobFormatReader.

        Args:
            file_path: Path to the blob file
            file_size: Size of the blob file in bytes
            selection: Optional list of blob indices to read (0-based)
        """
        self.file_path = file_path
        self.file_size = file_size
        self.blob_lengths: List[int] = []
        self.blob_offsets: List[int] = []
        self.returned = False

        # Read the file footer to get blob index
        self._read_index(selection)

    def _read_index(self, selection: Optional[List[int]] = None) -> None:
        """
        Read the blob index from the file footer.

        Args:
            selection: Optional list of blob indices to select
        """
        with open(self.file_path, 'rb') as f:
            # Read header (5 bytes from end: 4 bytes index length + 1 byte version)
            f.seek(self.file_size - 5)
            header = f.read(5)

            if len(header) != 5:
                raise IOError("Invalid blob file: cannot read header")

            # Parse header
            index_length = struct.unpack('<I', header[:4])[0]  # Little endian
            version = header[4]

            if version != 1:
                raise IOError(f"Unsupported blob file version: {version}")

            # Read index data
            f.seek(self.file_size - 5 - index_length)
            index_bytes = f.read(index_length)

            if len(index_bytes) != index_length:
                raise IOError("Invalid blob file: cannot read index")

            # Decompress blob lengths
            blob_lengths = DeltaVarintCompressor.decompress(index_bytes)

            # Calculate blob offsets
            blob_offsets = []
            offset = 0
            for length in blob_lengths:
                blob_offsets.append(offset)
                offset += length

            # Apply selection if provided
            if selection is not None:
                selected_lengths = []
                selected_offsets = []
                for idx in selection:
                    if 0 <= idx < len(blob_lengths):
                        selected_lengths.append(blob_lengths[idx])
                        selected_offsets.append(blob_offsets[idx])
                blob_lengths = selected_lengths
                blob_offsets = selected_offsets

            self.blob_lengths = blob_lengths
            self.blob_offsets = blob_offsets

    def read_batch(self) -> Optional['BlobRecordIterator']:
        """
        Read a batch of blob records.

        Returns:
            Iterator over blob records, or None if already returned
        """
        if self.returned:
            return None

        self.returned = True
        return BlobRecordIterator(self.file_path, self.blob_lengths, self.blob_offsets)

    def close(self) -> None:
        """Close the reader (no-op for file-based reader)."""
        pass


class BlobRecordIterator:
    """Iterator for blob records in a blob file."""

    def __init__(self, file_path: str, blob_lengths: List[int], blob_offsets: List[int]):
        """
        Initialize blob record iterator.

        Args:
            file_path: Path to the blob file
            blob_lengths: List of blob lengths
            blob_offsets: List of blob offsets
        """
        self.file_path = file_path
        self.blob_lengths = blob_lengths
        self.blob_offsets = blob_offsets
        self.current_position = 0

    def __iter__(self) -> Iterator[GenericRow]:
        """Return iterator."""
        return self

    def __next__(self) -> GenericRow:
        """
        Get next blob record.

        Returns:
            GenericRow containing a single blob field

        Raises:
            StopIteration: When no more records
        """
        if self.current_position >= len(self.blob_lengths):
            raise StopIteration

        # Create blob reference for the current blob
        # Skip magic number (4 bytes) and exclude length (8 bytes) + CRC (4 bytes) = 12 bytes
        blob_offset = self.blob_offsets[self.current_position] + 4  # Skip magic number
        blob_length = self.blob_lengths[self.current_position] - 16  # Exclude magic(4) + length(8) + CRC(4)

        # Create BlobDescriptor for this blob
        descriptor = BlobDescriptor(self.file_path, blob_offset, blob_length)
        blob = BlobRef(descriptor)

        self.current_position += 1

        # Return as GenericRow with single blob field
        from pypaimon.schema.data_types import DataField, AtomicType
        from pypaimon.table.row.row_kind import RowKind

        fields = [DataField(0, "blob_field", AtomicType("BLOB"))]
        return GenericRow([blob], fields, RowKind.INSERT)

    def returned_position(self) -> int:
        """Get current position in the iterator."""
        return self.current_position

    def file_path(self) -> str:
        """Get the file path."""
        return self.file_path

    def release_batch(self) -> None:
        """Release batch resources (no-op)."""
        pass
